address_restricted_ed25519.go

```diff
@@ -0,0 +1,241 @@
+	}
+
+	addr := Ed25519Address(redAddr.PubKeyHash)
+	return edSig.Valid(msg, &addr)
+}
+
+func (redAddr *RestrictedEd25519Address) Equal(other Address) bool {
+	otherAddr, is := other.(*RestrictedEd25519Address)
+	if !is {
+		return false
+	}
+
+	return redAddr.PubKeyHash == otherAddr.PubKeyHash &&
+		bytes.Equal(redAddr.Capabilities, otherAddr.Capabilities)
+}
+
+func (redAddr *RestrictedEd25519Address) Type() AddressType {
+	return AddressRestrictedEd25519
+}
+
+func (redAddr *RestrictedEd25519Address) Bech32(hrp NetworkPrefix) string {
+	return bech32String(hrp, redAddr)
+}
+
+func (redAddr *RestrictedEd25519Address) String() string {
+	return hexutil.EncodeHex(
+		byteutils.ConcatBytes(
+			redAddr.PubKeyHash[:],
+			[]byte{byte(len(redAddr.Capabilities))},
+			redAddr.Capabilities,
+		),
+	)
+}
+
+func (redAddr *RestrictedEd25519Address) Size() int {
+	return Ed25519AddressSerializedBytesSize +
+		serializer.SmallTypeDenotationByteSize +
+		len(redAddr.Capabilities)
+}
+
+func (redAddr *RestrictedEd25519Address) hasBit(bit int) bool {
+	byteIndex := bit / 8
+	if len(redAddr.Capabilities) <= byteIndex {
+		return false
+	}
+	bitIndex := bit % 8
+	return redAddr.Capabilities[byteIndex]&(1<<bitIndex) > 0
+}
+
+func (redAddr *RestrictedEd25519Address) setBit(bit int) {
+	byteIndex := bit / 8
+	for len(redAddr.Capabilities) <= byteIndex {
+		redAddr.Capabilities = append(redAddr.Capabilities, 0)
+	}
+	bitIndex := bit % 8
+	redAddr.Capabilities[byteIndex] |= 1 << bitIndex
+}
+
+func (redAddr *RestrictedEd25519Address) CanReceiveNativeTokens() bool {
+	return redAddr.hasBit(canReceiveNativeTokensBitIndex)
+}
+
+func (redAddr *RestrictedEd25519Address) CanReceiveMana() bool {
+	return redAddr.hasBit(canReceiveManaBitIndex)
+}
+
+func (redAddr *RestrictedEd25519Address) CanReceiveOutputsWithTimelockUnlockCondition() bool {
+	return redAddr.hasBit(canReceiveOutputsWithTimelockUnlockConditionBitIndex)
+}
+
+func (redAddr *RestrictedEd25519Address) CanReceiveOutputsWithExpirationUnlockCondition() bool {
+	return redAddr.hasBit(canReceiveOutputsWithExpirationUnlockConditionBitIndex)
+}
+
+func (redAddr *RestrictedEd25519Address) CanReceiveOutputsWithStorageDepositReturnUnlockCondition() bool {
+	return redAddr.hasBit(canReceiveOutputsWithStorageDepositReturnUnlockConditionBitIndex)
+}
+
+func (redAddr *RestrictedEd25519Address) CanReceiveAccountOutputs() bool {
+	return redAddr.hasBit(canReceiveAccountOutputsBitIndex)
+}
+
+func (redAddr *RestrictedEd25519Address) CanReceiveNFTOutputs() bool {
+	return redAddr.hasBit(canReceiveNFTOutputsBitIndex)
+}
+
+func (redAddr *RestrictedEd25519Address) CanReceiveDelegationOutputs() bool {
+	return redAddr.hasBit(canReceiveDelegationOutputsBitIndex)
+}
+
+// RestrictedEd25519AddressFromPubKey returns the address belonging to the given Ed25519 public key.
+func RestrictedEd25519AddressFromPubKey(pubKey ed25519.PublicKey,
```

Comment:
I think we should have another constructor from a pub key that does not take flags and sets them all to `0` by default.